### PR TITLE
Fixes to saving dates in assessment processor

### PR DESCRIPTION
**Pull request by gigxz**

### CODE REVIEW

File: drivers/hmis/spec/requests/hmis/submit_hud_assessments_spec.rb [hunk 1]
```diff
@@ -95,6 +123,71 @@
...
+        expect(errors).to be_empty
+        assessment = Hmis::Hud::Assessment.find(assessment_id)
+        expect(assessment).to be_present
+        expect(assessment.assessment_date).to eq(Date.parse(new_exit_date))
+        expect(assessment.assessment_detail.assessment_processor).to be_present
+
+        # entry and exit dates should NOT change on save
+        expect(assessment.enrollment.entry_date).to eq(enrollment_entry_date)
+        expect(assessment.enrollment.exit&.exit_date).to be_nil
+        # expect(assessment.enrollment.date_updated.inspect).to be == enrollment_date_updated.inspect
```
Comment (gigxz) [inline drivers/hmis/spec/requests/hmis/submit_hud_assessments_spec.rb]: I can't get this comparison working. The timestamps ARE the same, but for some reason the actual one is always rounded to fewer decimals:

```
       expected: == "Thu, 09 Feb 2023 16:19:51.791831287 EST -05:00"
            got:    "Thu, 09 Feb 2023 16:19:51.791831000 EST -05:00"
```
and
```
       expected: == "Thu, 09 Feb 2023 16:19:52.810008107 EST -05:00"
            got:    "Thu, 09 Feb 2023 16:19:52.810008000 EST -05:00"
```

File: drivers/hmis/app/graphql/mutations/submit_assessment.rb [hunk 2]
```diff
@@ -54,6 +54,9 @@ def resolve(input:)
       assessment_detail_valid = assessment.assessment_detail.valid?
 
       if assessment_valid && assessment_detail_valid
+        # call manually to save enrollment. figure out a better way
+        assessment.assessment_detail.assessment_processor.save!
```
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/submit_assessment.rb]: @eanders @awisema   here is my hacky solution:
This line, along with the `before_save :save_enrollment` hook in the processor, allows the Enrollment to save. I would love to figure out a way for the enrollment to save magically when the assessment_detail saves, like the rest of the records do.

File: drivers/hmis/app/models/hmis/form/assessment_processor.rb [hunk 3]
```diff
@@ -76,11 +81,15 @@ def enrollment_coc_factory(create: true)
...
-      personal_id: enrollment_factory.client.personal_id,
-      user_id: assessment_detail.assessment.user_id,
-      exit_date: assessment_detail.assessment.assessment_date,
-    )
+    if enrollment_factory.exit.present?
+      # Enrollment already has an Exit that's not tied to this processor (could occur in imported data..)
+      self.exit = enrollment_factory.exit
+    else
+      self.exit = enrollment_factory.build_exit(
+        personal_id: enrollment_factory.client.personal_id,
```
Comment (eanders) [inline drivers/hmis/app/models/hmis/form/assessment_processor.rb:89]: unless I'm mistaken, `enrollment_factory.personal_id` should be the same as the client's

File: drivers/hmis/spec/requests/hmis/submit_hud_assessments_spec.rb (hunk 1 above)
Comment (eanders) [reply drivers/hmis/spec/requests/hmis/submit_hud_assessments_spec.rb]: generally we compare these based on `to_date`, but you could `strftime` both sides, something like `strftime('%c')` to throw out the sub-second differences.

**Review by eanders — APPROVED** [review]

File: drivers/hmis/spec/requests/hmis/submit_hud_assessments_spec.rb (hunk 1 above)
Comment (gigxz) [reply drivers/hmis/spec/requests/hmis/submit_hud_assessments_spec.rb]: The timestamps are only milliseconds off from each other, so `%c` does not give enough specificity. Using `to_datetime` results in the same above error where the last 3 digits are zeros 🤔 

I got the tests passing by using `%Y-%m-%d %T.%3N`, just pushed that up.

**Comment by gigxz** [comment]
@awisema https://github.com/greenriver/hmis-warehouse/pull/2814/commits/88c61513e764310ad943c20e27b13a043f962505 updates the processor tests to test saving behavior

File: drivers/hmis/app/models/hmis/form/assessment_processor.rb (hunk 3 above)
Comment (gigxz) [reply drivers/hmis/app/models/hmis/form/assessment_processor.rb:89]: Yes, I think you're right. Just copying what was there

**Pull request merged**